Fixes #8396 "Postgres sync not respecting SSH tunneling"

dpsutton opened the pull request:
The ssh tunneling information is included in the pooled connection information so we reuse that. While diagnosing this, it seems that ssh tunneled connections will die after a couple minutes so i'm not sure how useful this has been in the meantime. I looked around and I couldn't find when the tunnels might time out on their own. One of the [associated issues](https://github.com/metabase/metabase/issues/10081) raises the possibility of some kind of heartbeat. I'm using `"idleConnectionTestPeriod"` to achieve this at the moment but the ticketer's concerns remain: we don't reestablish connections.

Fixes #8396

dpsutton committed:
Use db pool connection rather than raw details

connection pool is correctly setup with ssh tunneling information (if
any)

dpsutton committed:
Don't error when getting postgres enum types

this was throwing errors when we allowed our pooled connections to
close underneath us. But this particular step just isn't important
enough for that. If we don't have a connection available to us, let
the more robust query stuff fail it rather than this one off

dpsutton committed:
Set idleConnectionTestPeriod to 120 when ssh tunneling

The ssh tunnelling apparatus needs some work. The tunnels are created
and put in the spec details map but then this is tossed away when we
create the pool.

```clojure
(defn- create-pool!
  "Create a new C3P0 `ComboPooledDataSource` for connecting to the given `database`."
  [{:keys [id details], driver :engine, :as database}]
  {:pre [(map? database)]}
  (log/debug (u/format-color 'cyan (trs "Creating new connection pool for {0} database {1} ..." driver id)))
  (let [details-with-tunnel (ssh/include-ssh-tunnel details) ;; If the tunnel is disabled this returned unchanged
        spec                (connection-details->spec driver details-with-tunnel)
        properties          (data-warehouse-connection-pool-properties driver)]
    (connection-pool/connection-pool-spec spec (merge properties
                                                      (when (ssh/use-ssh-tunnel? details)
                                                        {"idleConnectionTestPeriod" 120})))))

```

this is swapped into an atom of {db_id -> pooleddatasource} but this
means we don't have the :tunnel-session and :tunnel-tracker info from
`start-ssh-tunnel`.

rlotun reviewed (commented):
A few comments and questions:

1. Please update the description to explicitly add text like "Fixes #8396" to automatically linked to the associated issue. This has the added bonus that the issue will be closed when this PR gets merged.

2. If this is a work-in-progress (WIP) PR please convert it to a draft while you work on it (see: https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft)

3. Are there supposed to be associated tests for this, or will that come later, because this is a WIP?

sbelak commented on src/metabase/driver/postgres.clj:
```diff
@@ -95,13 +95,17 @@
-                          "FROM pg_enum e "
-                          "LEFT JOIN pg_type t "
-                          "  ON t.oid = e.enumtypid")]))))
+  (try
```
In what cases will this fail?

dpsutton replied:
It has failed in the past by not acquiring a connection correctly _but_ the rest of the table syncing worked correctly. But checking for this set of enums would take down the entire table checking. I'm happy to back this try/catch out if that's preferable.

dpsutton replied:
It has failed in the past by not acquiring a connection correctly _but_ the rest of the table syncing worked correctly. But checking for this set of enums would take down the entire table checking. Happy to revert though if that's preferable.

sbelak replied:
I don't have a strong opinion on this, just trying to understand why (and potentially see if there are any further underlying issues). Perhaps add a comment for posterity.

dpsutton replied:
comment was in commit:

> this was throwing errors when we allowed our pooled connections to
close underneath us. But this particular step just isn't important
enough for that. If we don't have a connection available to us, let
the more robust query stuff fail it rather than this one off

can add explicit source comment as well.

dpsutton replied:
I've reverted the change so now its just using the correct connection. For two reasons: my code diff was large and it was complaining too much was untested and because phrasing the comment was getting long

dpsutton committed:
Remove connection pool keys when set to nil

our connection pool would end up with {db_id nil ...} for no
reason. Add tests to ensure.

This change is largely driven by the codox testing
requirement. There's no good way to test that the ssh tunneling
options includes

(when (ssh/use-ssh-tunnel? details) {"idleConnectionTestPeriod" 120})

and the test coverage checker was unhappy

dpsutton committed:
Correct ns form in test file

dpsutton commented:
>     1. Please update the description to explicitly add text like "Fixes #8396" to automatically linked to the associated issue. This has the added bonus that the issue will be closed when this PR gets merged.

thanks!

>     2. If this is a work-in-progress (WIP) PR please convert it to a draft while you work on it (see: https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request#converting-a-pull-request-to-a-draft)

It was until now I suppose. I'll make sure to use that judiciously in the future.

>     3. Are there supposed to be associated tests for this, or will that come later, because this is a WIP?

Testing this might be quite hard. Would require a server to tunnel to but the fix is so trivial. I have added tests for the function it calls (`sql-jdbc.conn/db->pooled-connection-spec` instead of `sql-jdbc.conn/connection-details->spec`) but the `enums` function this is in is unlikely to change in the future.

dpsutton commented:
rebased on master in hopes of fe and i18n tests passing

dpsutton committed:
Remove try/catch around enum types

there's already error handling around this sync step so let it
happen. The issue was that this check didn't use the correct
connection to respect tunnelling and therefore would throw an error in
getting enum types and fail the database sync step. That is now done
correctly so if this step fails the other steps would as well and we
no longer need to handle an error here specifically

robdaemon commented on src/metabase/driver/sql_jdbc/connection.clj:
```diff
@@ -87,7 +87,9 @@
-    (connection-pool/connection-pool-spec spec properties)))
+    (connection-pool/connection-pool-spec spec (merge properties
+                                                      (when (ssh/use-ssh-tunnel? details)
+                                                        {"idleConnectionTestPeriod" 120})))))
```
can this be done the ssh code itself? somewhere like here: https://github.com/metabase/metabase/blob/master/src/metabase/util/ssh.clj#L43-L59

This is a good fix - because the tunnels do timeout when there's no "heartbeat" or traffic, but with it here, MongoDB and other databases that don't derive from sql_jdbc won't benefit from it.

dpsutton replied:
I've made the change. As it stands it is hardcoded to 180 seconds. Will make a new ticket so this is user settable when configuring. But might be worth it to accept the current implementation.

robdaemon commented:
You're right that this won't entirely fix the problem - we don't reconnect when we lose the ssh connection. I don't think I introduced that when I swapped out our SSH library, but it's something that we definitely need to fix.

flamber commented:
@robdaemon That's existing issues https://github.com/metabase/metabase/issues?q=is%3Aissue+is%3Aopen+6114+6173

dpsutton committed:
Move ssh heartbeat into ssh.clj and not use jdbc

jdbc had a way to test idle connections but other non-jdbc datasources
obviously wouldn't benefit. Luckily our ssh connection can handle this
on its own

robdaemon commented on src/metabase/util/ssh.clj:
```diff
@@ -50,6 +54,9 @@
                                      (maybe-add-tunnel-private-key! tunnel-private-key tunnel-private-key-passphrase)
+                                     (.setSessionHeartbeat SessionHeartbeatController$HeartbeatType/IGNORE
+                                                           TimeUnit/SECONDS
+                                                           180)
```
added issue to track this becoming configurable

camsaul replied:
yeah normally at the very least it's nice to define this as a constant somewhere e.g.

```clj
(def ^:private heartbeat-interval-seconds 180)
```

which would let you override it with `with-redefs` or the like for test purposes. It's the whole "magic numbers" thing -- https://en.wikipedia.org/wiki/Magic_number_%28programming%29#Unnamed_numerical_constants

I think it's a minor issue tho so I don't want to block merge on this

vintas commented:
I've just tested out v0.37.0-rc2 docker image with rsa 4096 ssh connection and a postgres database on the other side of the connection.
Metabase is now able to sync the columns which didn't work on v0.36.6 but it loses this connection in around 10 minutes from the time I click on Save changes in the Add Database section.
The log says
`
[58b22ba2-5bf7-45f0-951a-c9e561edcad4] 2020-10-08T17:01:29+05:30 WARN org.apache.sshd.client.session.ClientSessionImpl ParameterizedMessage[messagePattern=exceptionCaught({})[state={}] {}: {}, stringArgs=[ClientSessionImpl[metabase@/x.x.x.x:22], Opened, InterruptedByTimeoutException, null], throwable=null]`

When I run any of the questions that I had saved during the first 10 minutes where it used to work it now gives me an error.

Logs:
`
[58b22ba2-5bf7-45f0-951a-c9e561edcad4] 2020-10-08T17:29:00+05:30 ERROR metabase.query-processor.middleware.catch-exceptions Error processing query: null
{:database_id 5,
 :started_at #t "2020-10-08T11:58:29.706010Z[GMT]",
 :via
 [{:status :failed,
   :class org.postgresql.util.PSQLException,
   :error "Connection to localhost:43727 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.",
   :stacktrace
   ["org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:280)"
    "org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)"
    "org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:195)"
    "org.postgresql.Driver.makeConnection(Driver.java:458)"
    "org.postgresql.Driver.connect(Driver.java:260)"
    "java.sql/java.sql.DriverManager.getConnection(Unknown Source)"
    "java.sql/java.sql.DriverManager.getConnection(Unknown Source)"
    "--> connection_pool$proxy_data_source$reify__20308.getConnection(connection_pool.clj:29)"],
   :state "08001"}
  {:status :failed,
   :class com.mchange.v2.resourcepool.CannotAcquireResourceException,
   :error "A ResourcePool could not acquire a resource from its primary factory or source.",
   :stacktrace
   ["com.mchange.v2.resourcepool.BasicResourcePool.awaitAvailable(BasicResourcePool.java:1507)"
    "com.mchange.v2.resourcepool.BasicResourcePool.prelimCheckoutResource(BasicResourcePool.java:644)"
    "com.mchange.v2.resourcepool.BasicResourcePool.checkoutResource(BasicResourcePool.java:554)"
    "com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutAndMarkConnectionInUse(C3P0PooledConnectionPool.java:758)"
    "com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutPooledConnection(C3P0PooledConnectionPool.java:685)"
    "com.mchange.v2.c3p0.impl.AbstractPoolBackedDataSource.getConnection(AbstractPoolBackedDataSource.java:140)"
    "--> driver.sql_jdbc.execute$fn__75684.invokeStatic(execute.clj:157)"
    "driver.sql_jdbc.execute$fn__75684.invoke(execute.clj:155)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:386)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
    "driver.sql_jdbc$fn__77111.invokeStatic(sql_jdbc.clj:49)"
    "driver.sql_jdbc$fn__77111.invoke(sql_jdbc.clj:47)"
    "query_processor.context$executef.invokeStatic(context.clj:59)"
    "query_processor.context$executef.invoke(context.clj:48)"
    "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
    "query_processor.context.default$default_runf.invoke(default.clj:67)"
    "query_processor.context$runf.invokeStatic(context.clj:45)"
    "query_processor.context$runf.invoke(context.clj:39)"
    "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
    "query_processor.reducible$pivot.invoke(reducible.clj:31)"
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45710.invoke(mbql_to_native.clj:26)"
    "query_processor.middleware.check_features$check_features$fn__44986.invoke(check_features.clj:42)"
    "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__45875.invoke(optimize_datetime_filters.clj:133)"
    "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__43792.invoke(auto_parse_filter_values.clj:44)"
    "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47404.invoke(wrap_value_literals.clj:142)"
    "query_processor.middleware.annotate$add_column_info$fn__43554.invoke(annotate.clj:574)"
    "query_processor.middleware.permissions$check_query_permissions$fn__44861.invoke(permissions.clj:64)"
    "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46393.invoke(pre_alias_aggregations.clj:40)"
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45059.invoke(cumulative_aggregations.clj:61)"
    "query_processor.middleware.resolve_joins$resolve_joins$fn__46925.invoke(resolve_joins.clj:183)"
    "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39279.invoke(add_implicit_joins.clj:245)"
    "query_processor.middleware.large_int_id$convert_id_to_string$fn__45671.invoke(large_int_id.clj:44)"
    "query_processor.middleware.limit$limit$fn__45696.invoke(limit.clj:38)"
    "query_processor.middleware.format_rows$format_rows$fn__45651.invoke(format_rows.clj:81)"
    "query_processor.middleware.desugar$desugar$fn__45125.invoke(desugar.clj:22)"
    "query_processor.middleware.binning$update_binning_strategy$fn__44151.invoke(binning.clj:229)"
    "query_processor.middleware.resolve_fields$resolve_fields$fn__44667.invoke(resolve_fields.clj:24)"
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__38828.invoke(add_dimension_projections.clj:316)"
    "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__39035.invoke(add_implicit_clauses.clj:141)"
    "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39428.invoke(add_source_metadata.clj:105)"
    "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46590.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
    "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43739.invoke(auto_bucket_datetimes.clj:125)"
    "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44714.invoke(resolve_source_table.clj:46)"
    "query_processor.middleware.parameters$substitute_parameters$fn__46375.invoke(parameters.clj:114)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44766.invoke(resolve_referenced.clj:80)"
    "query_processor.middleware.expand_macros$expand_macros$fn__45381.invoke(expand_macros.clj:158)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39459.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47287.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46601$fn__46605.invoke(resolve_database_and_driver.clj:33)"
    "driver$do_with_driver.invokeStatic(driver.clj:61)"
    "driver$do_with_driver.invoke(driver.clj:57)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46601.invoke(resolve_database_and_driver.clj:27)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45599.invoke(fetch_source_query.clj:267)"
    "query_processor.middleware.store$initialize_store$fn__47296$fn__47297.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
    "query_processor.store$do_with_store.invoke(store.clj:40)"
    "query_processor.middleware.store$initialize_store$fn__47296.invoke(store.clj:10)"
    "query_processor.middleware.cache$maybe_return_cached_results$fn__44643.invoke(cache.clj:209)"
    "query_processor.middleware.validate$validate_query$fn__47305.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__45723.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39297.invoke(add_rows_truncated.clj:36)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47272.invoke(results_metadata.clj:147)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__45002.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__46464.invoke(process_userland_query.clj:136)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__44945.invoke(catch_exceptions.clj:174)"
    "query_processor.reducible$async_qp$qp_STAR___38091$thunk__38092.invoke(reducible.clj:101)"
    "query_processor.reducible$async_qp$qp_STAR___38091$fn__38094.invoke(reducible.clj:106)"]}
  {:status :failed,
   :class java.sql.SQLException,
   :error "Connections could not be acquired from the underlying database!",
   :stacktrace
   ["com.mchange.v2.sql.SqlUtils.toSQLException(SqlUtils.java:118)"
    "com.mchange.v2.c3p0.impl.C3P0PooledConnectionPool.checkoutPooledConnection(C3P0PooledConnectionPool.java:692)"
    "com.mchange.v2.c3p0.impl.AbstractPoolBackedDataSource.getConnection(AbstractPoolBackedDataSource.java:140)"
    "--> driver.sql_jdbc.execute$fn__75684.invokeStatic(execute.clj:157)"
    "driver.sql_jdbc.execute$fn__75684.invoke(execute.clj:155)"
    "driver.sql_jdbc.execute$execute_reducible_query.invokeStatic(execute.clj:386)"
    "driver.sql_jdbc.execute$execute_reducible_query.invoke(execute.clj:377)"
    "driver.sql_jdbc$fn__77111.invokeStatic(sql_jdbc.clj:49)"
    "driver.sql_jdbc$fn__77111.invoke(sql_jdbc.clj:47)"
    "query_processor.context$executef.invokeStatic(context.clj:59)"
    "query_processor.context$executef.invoke(context.clj:48)"
    "query_processor.context.default$default_runf.invokeStatic(default.clj:69)"
    "query_processor.context.default$default_runf.invoke(default.clj:67)"
    "query_processor.context$runf.invokeStatic(context.clj:45)"
    "query_processor.context$runf.invoke(context.clj:39)"
    "query_processor.reducible$pivot.invokeStatic(reducible.clj:34)"
    "query_processor.reducible$pivot.invoke(reducible.clj:31)"
    "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__45710.invoke(mbql_to_native.clj:26)"
    "query_processor.middleware.check_features$check_features$fn__44986.invoke(check_features.clj:42)"
    "query_processor.middleware.optimize_datetime_filters$optimize_datetime_filters$fn__45875.invoke(optimize_datetime_filters.clj:133)"
    "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__43792.invoke(auto_parse_filter_values.clj:44)"
    "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__47404.invoke(wrap_value_literals.clj:142)"
    "query_processor.middleware.annotate$add_column_info$fn__43554.invoke(annotate.clj:574)"
    "query_processor.middleware.permissions$check_query_permissions$fn__44861.invoke(permissions.clj:64)"
    "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__46393.invoke(pre_alias_aggregations.clj:40)"
    "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__45059.invoke(cumulative_aggregations.clj:61)"
    "query_processor.middleware.resolve_joins$resolve_joins$fn__46925.invoke(resolve_joins.clj:183)"
    "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__39279.invoke(add_implicit_joins.clj:245)"
    "query_processor.middleware.large_int_id$convert_id_to_string$fn__45671.invoke(large_int_id.clj:44)"
    "query_processor.middleware.limit$limit$fn__45696.invoke(limit.clj:38)"
    "query_processor.middleware.format_rows$format_rows$fn__45651.invoke(format_rows.clj:81)"
    "query_processor.middleware.desugar$desugar$fn__45125.invoke(desugar.clj:22)"
    "query_processor.middleware.binning$update_binning_strategy$fn__44151.invoke(binning.clj:229)"
    "query_processor.middleware.resolve_fields$resolve_fields$fn__44667.invoke(resolve_fields.clj:24)"
    "query_processor.middleware.add_dimension_projections$add_remapping$fn__38828.invoke(add_dimension_projections.clj:316)"
    "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__39035.invoke(add_implicit_clauses.clj:141)"
    "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__39428.invoke(add_source_metadata.clj:105)"
    "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__46590.invoke(reconcile_breakout_and_order_by_bucketing.clj:98)"
    "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__43739.invoke(auto_bucket_datetimes.clj:125)"
    "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__44714.invoke(resolve_source_table.clj:46)"
    "query_processor.middleware.parameters$substitute_parameters$fn__46375.invoke(parameters.clj:114)"
    "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__44766.invoke(resolve_referenced.clj:80)"
    "query_processor.middleware.expand_macros$expand_macros$fn__45381.invoke(expand_macros.clj:158)"
    "query_processor.middleware.add_timezone_info$add_timezone_info$fn__39459.invoke(add_timezone_info.clj:15)"
    "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__47287.invoke(splice_params_in_response.clj:32)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46601$fn__46605.invoke(resolve_database_and_driver.clj:33)"
    "driver$do_with_driver.invokeStatic(driver.clj:61)"
    "driver$do_with_driver.invoke(driver.clj:57)"
    "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__46601.invoke(resolve_database_and_driver.clj:27)"
    "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__45599.invoke(fetch_source_query.clj:267)"
    "query_processor.middleware.store$initialize_store$fn__47296$fn__47297.invoke(store.clj:11)"
    "query_processor.store$do_with_store.invokeStatic(store.clj:46)"
    "query_processor.store$do_with_store.invoke(store.clj:40)"
    "query_processor.middleware.store$initialize_store$fn__47296.invoke(store.clj:10)"
    "query_processor.middleware.cache$maybe_return_cached_results$fn__44643.invoke(cache.clj:209)"
    "query_processor.middleware.validate$validate_query$fn__47305.invoke(validate.clj:10)"
    "query_processor.middleware.normalize_query$normalize$fn__45723.invoke(normalize_query.clj:22)"
    "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__39297.invoke(add_rows_truncated.clj:36)"
    "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__47272.invoke(results_metadata.clj:147)"
    "query_processor.middleware.constraints$add_default_userland_constraints$fn__45002.invoke(constraints.clj:42)"
    "query_processor.middleware.process_userland_query$process_userland_query$fn__46464.invoke(process_userland_query.clj:136)"
    "query_processor.middleware.catch_exceptions$catch_exceptions$fn__44945.invoke(catch_exceptions.clj:174)"
    "query_processor.reducible$async_qp$qp_STAR___38091$thunk__38092.invoke(reducible.clj:101)"
    "query_processor.reducible$async_qp$qp_STAR___38091$fn__38094.invoke(reducible.clj:106)"],
   :state nil}],
 :json_query
 {:constraints {:max-results 10000, :max-results-bare-rows 2000},
  :type :native,
  :middleware {:js-int-to-string? true},
  :native
  {:query
   "select\n\tcred_gen_date,\n\tcred_note_no,\n\tcred_narration,\n\tcred_is_confirm,\n\tcred_gen_created_datetime,\n\tcred_print_status,\n\tusername,\n\tiv_invoice_no,\n\tiv_title,\n\tiv_customer_name,\n\tiv_narration,\n\tiv_gen_date\nfrom\n\tinvoice_cred_note_generation\nleft join auth_user on\n\tcred_created_emply_id = id\nleft join invoice_iv_generation on\n\tcred_iv_gen_id = iv_gen_id\nwhere\n\tcred_gen_enabled = 'Y'\n\tand cred_gen_deleted = 'N'\norder by\n\tcred_gen_date desc,\n\tcred_gen_id desc",
   :template-tags {}},
  :database 5,
  :parameters [],
  :async? true,
  :cache-ttl nil},
 :status :failed,
 :class java.net.ConnectException,
 :stacktrace
 ["java.base/java.net.PlainSocketImpl.socketConnect(Native Method)"
  "java.base/java.net.AbstractPlainSocketImpl.doConnect(Unknown Source)"
  "java.base/java.net.AbstractPlainSocketImpl.connectToAddress(Unknown Source)"
  "java.base/java.net.AbstractPlainSocketImpl.connect(Unknown Source)"
  "java.base/java.net.SocksSocketImpl.connect(Unknown Source)"
  "java.base/java.net.Socket.connect(Unknown Source)"
  "org.postgresql.core.PGStream.<init>(PGStream.java:75)"
  "org.postgresql.core.v3.ConnectionFactoryImpl.tryConnect(ConnectionFactoryImpl.java:91)"
  "org.postgresql.core.v3.ConnectionFactoryImpl.openConnectionImpl(ConnectionFactoryImpl.java:192)"
  "org.postgresql.core.ConnectionFactory.openConnection(ConnectionFactory.java:49)"
  "org.postgresql.jdbc.PgConnection.<init>(PgConnection.java:195)"
  "org.postgresql.Driver.makeConnection(Driver.java:458)"
  "org.postgresql.Driver.connect(Driver.java:260)"
  "java.sql/java.sql.DriverManager.getConnection(Unknown Source)"
  "java.sql/java.sql.DriverManager.getConnection(Unknown Source)"
  "--> connection_pool$proxy_data_source$reify__20308.getConnection(connection_pool.clj:29)"],
 :context :question,
 :error "Connection to localhost:43727 refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections.",
 :row_count 0,
 :running_time 0,
 :data {:rows [], :cols []}}
`

Would this be considered a separate issue?

flamber commented:
@vintas Thanks for testing :tada: Yes, that would be a separate issue, but likely one of the existing: #3941 #6114 #10081 #13399
